### PR TITLE
feat: Adds config modifier to bigquery job options

### PR DESCRIPTION
**Pull request by hemanshv**
Implement Configuration Modifier in BigQuery Jobs.
**Note:** Integration tests are not added intentionally as nothing is going to change the way backend apis are treated.

### CODE REVIEW

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs [hunk 1]
```diff
@@ -43,6 +44,12 @@ public sealed class CreateCopyJobOptions : JobCreationOptions
         /// </summary>
         public CopyOperationType? OperationType { get; set; }
 
+        /// <summary>
```
Comment (Rishabh-V) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs:47]: (As a person who doesn't have a context) This documentation doesn't help me understand what is its purpose.  Helper?

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ namespace Google.Cloud.BigQuery.V2
     /// Options for <c>CreateExtractJob</c> operations.
     /// </summary>
     public sealed class CreateExtractJobOptions : JobCreationOptions
-    {
+    { 
         /// <summary>
```
Comment (Rishabh-V) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs:25]: nit: Adds unnecessary whitespace.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs [hunk 3]
```diff
@@ -52,6 +52,12 @@ public sealed class CreateExtractJobOptions : JobCreationOptions
         /// </summary>
         public bool? UseAvroLogicalTypes { get; set; }
 
+        /// <summary>
+        /// A helper to modify the underlying job configuration.
+        /// </summary>
+        public Action<JobConfigurationExtract> ConfigurationModifier { get; set; }
+
```
Comment (Rishabh-V) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs:59]: Same as above and if its applicable in other places.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs [hunk 4]
```diff
@@ -78,6 +84,10 @@ internal void ModifyRequest(JobConfigurationExtract extract)
             {
                 extract.UseAvroLogicalTypes = UseAvroLogicalTypes;
             }
+            if(ConfigurationModifier != null)
```
Comment (Rishabh-V) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs:87]: Here as well.

**Review by Rishabh-V — COMMENTED** [review]
Just had a quick look (as it is end of the day) at the production code. I have shared a few minor formatting nits. I will have another pass tomorrow after going through the document. Thanks.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs (hunk 1 above)
Comment (hemanshv) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs:47]: Updated

**Comment by hemanshv** [comment]
Hi @jskeet and @amanda-tarafa ,
Please have a look and share your feedback.

Thanks,

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/CreateCopyJobOptionsTest.cs [hunk 5]
```diff
@@ -27,13 +27,15 @@ public void ModifyRequest()
                 CreateDisposition = CreateDisposition.CreateIfNeeded,
                 WriteDisposition = WriteDisposition.WriteIfEmpty,
                 DestinationEncryptionConfiguration = new EncryptionConfiguration { KmsKeyName = "projects/1/locations/us/keyRings/1/cryptoKeys/1" },
-                OperationType = CopyOperationType.Clone
+                OperationType = CopyOperationType.Clone,
+                ConfigurationModifier = (options) => options.ETag = "test"
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/CreateCopyJobOptionsTest.cs:31]: Nit: there's no need for bracket round the lambda parameter, and it's non-idiomatic IMO. So just `options => options.ETag = "test"` would be better.

Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/CreateCopyJobOptionsTest.cs:31]: (Same for all the other examples, I'm afraid...)

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs (hunk 2 above)
Comment (jskeet) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateExtractJobOptions.cs:25]: Whitespace still appears to be there...

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs [hunk 6]
```diff
@@ -43,6 +44,11 @@ public sealed class CreateCopyJobOptions : JobCreationOptions
...
+        /// Delegate that is used to modify the underlying <see cref="JobConfigurationTableCopy"/>.
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs:48]: Let's get everything else in the PR ready, then I'll have a go at wordsmithing this. (I owe the team a doc about XML documentation comments...)

**Review by jskeet — COMMENTED** [review]
Generally happy with the shape - a couple of C# idiomatic nits (that unfortunately mean changes across multiple files). Once they're in, I can work on the docs a bit more - or that could even be after this PR is merged.

I note that if we apply the same pattern (with different naming) to other operations (particularly query) then #11009 might become obsolete - or at least, we could defer that until questions about whether the approach is going to spread to other operations has been answered.

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/CreateCopyJobOptionsTest.cs (hunk 5 above)
Comment (hemanshv) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Tests/CreateCopyJobOptionsTest.cs]: done

File: apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs [hunk 7]
```diff
@@ -61,6 +67,10 @@ internal void ModifyRequest(JobConfigurationTableCopy copy)
             {
                 copy.OperationType = EnumMap.ToApiValue(OperationType.Value);
             }
+            if (ConfigurationModifier != null)
```
Comment (jskeet) [inline apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs:70]: All of these should be `ConfigurationModifier?.Invoke(copy);` for brevity.

I *thought* I'd added a comment to that effect before, but I now can't see it...

Comment (hemanshv) [reply apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/CreateCopyJobOptions.cs]: done

**Review by jskeet — APPROVED** [review]
Looks good, other than the XML element. I'll create a PR for that after you've merged.
